src/libs/actions/OnyxUpdateManager/utils/index.ts

```diff
@@ -0,0 +1,137 @@
+            // so that we can continue searching for the next update after all gaps
+            gapExists = true;
+            firstUpdateAfterGaps = undefined;
+
+            // If there is a gap, it means the previous update is the latest missing update.
+            latestMissingUpdateID = Number(update.previousUpdateID);
+        }
+    }
+
+    // When "firstUpdateAfterGaps" is not set yet, we need to set it to the last update in the list,
+    // because we will fetch all missing updates up to the previous one and can then always apply the last update in the deferred updates.
+    const firstUpdateAfterGapWithFallback = firstUpdateAfterGaps ?? Number(updateValues[updateValues.length - 1].lastUpdateID);
+
+    let updatesAfterGaps: DeferredUpdatesDictionary = {};
+    if (gapExists) {
+        updatesAfterGaps = Object.entries(updates).reduce<DeferredUpdatesDictionary>(
+            (accUpdates, [lastUpdateID, update]) => ({
+                ...accUpdates,
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                ...(Number(lastUpdateID) >= firstUpdateAfterGapWithFallback ? {[Number(lastUpdateID)]: update} : {}),
+            }),
+            {},
+        );
+    }
+
+    return {applicableUpdates, updatesAfterGaps, latestMissingUpdateID};
+}
+
+// This function will check for gaps in the deferred updates and
+// apply the updates in order after the missing updates are fetched and applied
+function validateAndApplyDeferredUpdates(clientLastUpdateID?: number): Promise<void> {
+    const lastUpdateIDFromClient = clientLastUpdateID ?? lastUpdateIDAppliedToClient ?? 0;
+
+    // We only want to apply deferred updates that are newer than the last update that was applied to the client.
+    // At this point, the missing updates from "GetMissingOnyxUpdates" have been applied already, so we can safely filter out.
+    const pendingDeferredUpdates = Object.entries(deferredUpdatesProxy.deferredUpdates).reduce<DeferredUpdatesDictionary>(
+        (accUpdates, [lastUpdateID, update]) => ({
+            ...accUpdates,
+            ...(Number(lastUpdateID) > lastUpdateIDFromClient ? {[Number(lastUpdateID)]: update} : {}),
+        }),
+        {},
+    );
+
+    // If there are no remaining deferred updates after filtering out outdated ones,
+    // we can just unpause the queue and return
+    if (Object.values(pendingDeferredUpdates).length === 0) {
+        return Promise.resolve();
+    }
+
+    const {applicableUpdates, updatesAfterGaps, latestMissingUpdateID} = detectGapsAndSplit(pendingDeferredUpdates, lastUpdateIDFromClient);
+
+    // If we detected a gap in the deferred updates, only apply the deferred updates before the gap,
+    // re-fetch the missing updates and then apply the remaining deferred updates after the gap
+    if (latestMissingUpdateID) {
+        return new Promise((resolve, reject) => {
+            deferredUpdatesProxy.deferredUpdates = {};
```

Comment:
I think we should move this line to right after calling the synchronous method `detectGapsAndSplit`.

Reply:
Just note for myself
> The executor is called synchronously (as soon as the Promise is constructed) with the resolveFunc and rejectFunc functions as arguments.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise

We also early return when existingDeferredUpdatesCount > 0

So there won't be an issue that updates are discarded without being applied.